lib/jsonrpc.ml

```diff
@@ -148,407 +142,32 @@ let a_of_response ?(id=Int 0L) ?(version=V1) ~empty ~append response =
-       | ',' when is_parsing_object s -> s.cursor <- Expect_object_key
-       | ',' -> s.cursor <- Expect_value
-       | '}' when is_parsing_object s -> pop_stack s
-       | '}' -> raise_unexpected_char s c "comma_or_end"
-       | ']' when not (is_parsing_object s) -> pop_stack s
-       | ']' -> raise_unexpected_char s c "comma_or_end"
-       | _ when is_space c -> update_line_num s c
-       | _ -> raise_unexpected_char s c "comma_or_end")
-
-    | Expect_object_key ->
-      (match c with
-       | '"' ->
-         (match s.stack with
-          | IObject fields :: tl -> s.stack <- IObject_needs_key fields :: tl
-          | io :: _ -> raise_internal_error s ("unexpected " ^ (ivalue_to_str io) ^ " on stack at object_key")
-          | [] -> raise_internal_error s "empty stack at object_key");
-         s.cursor <- In_string []
-       | _ when is_space c -> update_line_num s c
-       | _ -> raise_unexpected_char s c "object_key")
-
-    | Done _ -> raise_internal_error s "parse called when parse_state is 'Done'"
-
-  type parse_result =
-    | Json_value of t
-    | Json_parse_incomplete of parse_state
-
-  let parse state str =
-    while get_parse_result state = None do
-      parse_char state (str ());
-      (* This is here instead of inside parse_char since
-                                           parse_char makes (tail-)recursive calls without
-                                           consuming a character.
-      *)
-      state.num_chars_parsed <- state.num_chars_parsed + 1;
-    done;
-    match get_parse_result state with
-    | Some v -> Json_value v
-    | None -> Json_parse_incomplete state
-
-  (* This is really only required for numbers, since they are only
-           terminated by whitespace, but end-of-file or end-of-connection
-           qualifies as whitespace.
-
-           The parser might also be just eating whitespace, expecting the
-           start of a json value.
-  *)
-  let finish_parse state =
-    match parse state (fun () -> ' ') with
-    | Json_value v -> Some v
-    | Json_parse_incomplete _ ->
-      if state.cursor = Start then None
-      else raise_unterminated_value state (current_cursor_value state.cursor)
-
-  let num_chars_parsed state = state.num_chars_parsed
-
-  let of_stream str =
-    match parse (init_parse_state ()) str with
-    | Json_value v -> v
-    | Json_parse_incomplete st ->
-      match finish_parse st with
-      | Some v -> v
-      | None -> raise_unterminated_value st (current_cursor_value st.cursor)
-
-  let of_string str =
-    let i = ref (-1) in
-    let next () =
-      incr i;
-      str.[ !i ] in
-    of_stream next
-end
-
-let of_fct = Parser.of_stream
-
-let of_string = Parser.of_string
+let of_string s = s |> Y.from_string |> json_to_rpc
 let of_a ~next_char b =
-  Parser.of_stream (fun () -> next_char b)
+  let buf = Buffer.create 2048 in
+  let rec acc () =
+    try 
+      Buffer.add_char buf (next_char b);
+      acc ()
+    with _ -> ()
```

Comment:
I'm a bit uncomfortable matching all exceptions here. This can swallow an unexpected exception.

Reply:
Then we will fail parsing. Unfortunately I have no control on `next_char`, so I don't know how it can explode and I don't want to break backward compatibility too much. What would you suggest?

Reply:
I have kept the unnice function but deprecated it